transports/ftp/src/test/java/org/mule/transport/ftp/FtpReconnectionNotifierTestCase.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.ftpserver.usermanager.PropertiesUserManagerFactory;
+import org.apache.ftpserver.usermanager.impl.BaseUser;
+import org.apache.ftpserver.usermanager.impl.WritePermission;
+import org.junit.Rule;
+import org.junit.Test;
+
+
+public class FtpReconnectionNotifierTestCase extends FunctionalTestCase
+{
+
+    private final Integer RECONNECTION_ATTEMPTS = 2;
+    private final Integer POLLING_FREQUENCY = 100;
+
+    @Rule
+    public SystemProperty countSystemProperty = new SystemProperty("count", RECONNECTION_ATTEMPTS.toString());
+
+    @Rule
+    public SystemProperty pollingFrequencySystemProperty = new SystemProperty("pollingFrequency", POLLING_FREQUENCY.toString());
+
+    private TestReconnectNotifier mockNotifier = null;
+
+
+    protected String getConfigFile()
+    {
+        return "ftp-reconnection-notifier-fails-config.xml";
+    }
+
+    @Test
+    public void testNotificationOnReconnectionAttemptFailure() throws Exception
+    {
+        mockNotifier = mock(TestReconnectNotifier.class);
+        Connector c = muleContext.getRegistry().lookupConnector("FTP");
+        assertNotNull(c);
+        RetryPolicyTemplate rpf = c.getRetryPolicyTemplate();
+        assertNotNull(rpf);
+        rpf.setNotifier(mockNotifier);
+        MuleFtpServer server = new MuleFtpServer(2221, 1);
+        server.start();
+        testServerConnected();
+        server.stop();
+        testServerDisconnected();
+    }
+
+    private void verifyOnSuccessNotification()
+    {
+        PollingProber pollingProber = new PollingProber(POLLING_FREQUENCY * 5, POLLING_FREQUENCY);
+        pollingProber.check(new JUnitProbe()
+        {
+            @Override
+            protected boolean test() throws Exception
+            {
+                verify(mockNotifier, times(1)).onSuccess(any(RetryContext.class));
+                return true;
+            }
+
+            @Override
+            public String describeFailure()
+            {
+                return "Success notification wasn't triggered";
+            }
+        });
+    }
+
+    private void verifyOnFailureNotification()
+    {
+        PollingProber pollingProber = new PollingProber(POLLING_FREQUENCY * 5, POLLING_FREQUENCY);
+        pollingProber.check(new JUnitProbe()
+        {
+            @Override
+            protected boolean test() throws Exception
+            {
+                verify(mockNotifier, atLeast(1)).onFailure(any(RetryContext.class), any(Throwable.class));
+                return true;
+            }
+
+            @Override
+            public String describeFailure()
+            {
+                return "Failure notification wasn't triggered";
+            }
+        });
+    }
+
+    private void testServerConnected() throws Exception
+    {
+        runFlow("main-test");
+        verifyOnSuccessNotification();
+    }
+
+    private void testServerDisconnected()
+    {
+        try
+        {
+            runFlow("main-test");
```

Comment:
Well If reconnection fails an exception is triggered. 
If I don't throw the expcetion, I can't compare the test results.